File: app/src/main/java/org/isoron/uhabits/widgets/HistoryWidget.java
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.widgets;

import android.app.*;
import android.content.*;
import android.support.annotation.*;
import android.view.*;

import org.isoron.uhabits.activities.common.views.*;
import org.isoron.uhabits.models.*;
import org.isoron.uhabits.utils.*;
import org.isoron.uhabits.widgets.views.*;

public class HistoryWidget extends BaseWidget
{
    @NonNull
    private Habit habit;

    public HistoryWidget(@NonNull Context context, int id, @NonNull Habit habit)
    {
        super(context, id);
        this.habit = habit;
    }

    @Override
    public PendingIntent getOnClickPendingIntent(Context context)
    {
        return pendingIntentFactory.showHabit(habit);
    }

    @Override
    public void refreshData(View view)
    {
        GraphWidgetView widgetView = (GraphWidgetView) view;
        HistoryChart chart = (HistoryChart) widgetView.getDataView();

        int color = ColorUtils.getColor(getContext(), habit.getColor());
        int[] values = habit.getCheckmarks().getAllValues();

        chart.setColor(color);
        chart.setCheckmarks(values);
    }

    @Override
    protected View buildView()
    {
        HistoryChart dataView = new HistoryChart(getContext());
        GraphWidgetView widgetView =
            new GraphWidgetView(getContext(), dataView);
        widgetView.setTitle(habit.getName());
        return widgetView;
    }

    @Override
    protected int getDefaultHeight()
    {
        return 250;
    }

    @Override
    protected int getDefaultWidth()
    {
        return 250;
    }
}

File: android/uhabits-android/src/main/java/org/isoron/uhabits/widgets/HistoryWidget.kt
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.widgets

import android.app.PendingIntent
import android.content.Context
import android.view.View
import org.isoron.uhabits.activities.common.views.HistoryChart
import org.isoron.uhabits.core.models.Habit
import org.isoron.uhabits.utils.toThemedAndroidColor
import org.isoron.uhabits.widgets.views.GraphWidgetView

class HistoryWidget(
    context: Context,
    id: Int,
    private val habit: Habit,
    private val firstWeekday: Int
) : BaseWidget(context, id) {

    override fun getOnClickPendingIntent(context: Context): PendingIntent {
        return pendingIntentFactory.showHabit(habit)
    }

    override fun refreshData(view: View) {
        val widgetView = view as GraphWidgetView
        widgetView.setBackgroundAlpha(preferedBackgroundAlpha)
        if (preferedBackgroundAlpha >= 255) widgetView.setShadowAlpha(0x4f)
        (widgetView.dataView as HistoryChart).apply {
            setFirstWeekday(firstWeekday)
            setSkipEnabled(prefs.isSkipEnabled)
            setColor(habit.color.toThemedAndroidColor(context))
            setEntries(habit.computedEntries.getAllValues())
            setNumerical(habit.isNumerical)
        }
    }

    override fun buildView() =
        GraphWidgetView(context, HistoryChart(context)).apply {
            setTitle(habit.name)
        }

    override fun getDefaultHeight() = 250
    override fun getDefaultWidth() = 250
}
